File: mobile/android/app/src/main/java/com/example/seniordesignmobileapp/viewmodel/AggregatorViewModel.kt
package com.example.seniordesignmobileapp.viewmodel

import android.content.Context
import android.os.SystemClock
import androidx.lifecycle.ViewModel
import androidx.lifecycle.ViewModelProvider
import androidx.lifecycle.viewModelScope
import androidx.lifecycle.viewmodel.initializer
import androidx.lifecycle.viewmodel.viewModelFactory
import com.example.seniordesignmobileapp.ble.BleAggregatorController
import com.example.seniordesignmobileapp.domain.PostureAnalysisCoordinator
import com.example.seniordesignmobileapp.model.AggregatorUiState
import com.example.seniordesignmobileapp.model.AnalysisUiState
import com.example.seniordesignmobileapp.model.ModeledGravityVector
import com.example.seniordesignmobileapp.model.ModeledNodeUiState
import com.example.seniordesignmobileapp.model.ModelingUiState
import com.example.seniordesignmobileapp.model.SavedSessionSummary
import com.example.seniordesignmobileapp.modeling.NodePoseMath
import com.example.seniordesignmobileapp.recording.SessionRecorder
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.SharingStarted
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.combine
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.flow.stateIn
import kotlinx.coroutines.launch
import kotlinx.coroutines.runBlocking

class AggregatorViewModel(
    applicationContext: Context,
) : ViewModel() {
    private val controller = BleAggregatorController(applicationContext)
    private val sessionRecorder = SessionRecorder(applicationContext)
    private val recordingState = MutableStateFlow(RecordingState())
    private val analysisCoordinator = PostureAnalysisCoordinator()
    private val analysisState = MutableStateFlow(AnalysisUiState())
    private val modelingState = MutableStateFlow(ModelingUiState())
    private val latestSamplesBySensor = linkedMapOf<Int, TimedSensorSample>()
    private var lastRecordedSampleReceivedAtElapsedMs: Long? = null
    private var lastRecordedStatusReceivedAtElapsedMs: Long? = null
    private var lastAnalyzedSampleReceivedAtElapsedMs: Long? = null
    private var lastAnalyzedStatusReceivedAtElapsedMs: Long? = null

    val uiState: StateFlow<AggregatorUiState> = combine(
        controller.uiState,
        recordingState,
        analysisState,
        modelingState,
    ) { bleState, recording, analysis, modeling ->
        bleState.copy(
            isRecording = recording.isRecording,
            recordingSessionName = recording.sessionName,
            recordingSessionPath = recording.sessionPath,
            recordingStartedAtElapsedMs = recording.startedAtElapsedMs,
            recordedSampleCount = recording.recordedSampleCount,
            recordedStatusCount = recording.recordedStatusCount,
            savedSessions = recording.savedSessions,
            recordingErrorMessage = recording.errorMessage,
            analysis = analysis,
            modeling = modeling,
        )
    }.stateIn(
        scope = viewModelScope,
        started = SharingStarted.WhileSubscribed(stopTimeoutMillis = 5_000),
        initialValue = controller.uiState.value,
    )

    init {
        viewModelScope.launch {
            refreshSavedSessions()
        }
        viewModelScope.launch {
            controller.uiState.collect { bleState ->
                recordIfNeeded(bleState)
                analyzeIfNeeded(bleState)
                updateModelingState(bleState)
            }
        }
    }

    fun onBlePermissionsChanged(granted: Boolean) {
        if (granted) {
            controller.start()
        } else {
            resetAnalysis()
            controller.stop()
        }
    }

    fun reconnect() {
        resetAnalysis()
        controller.restart()
    }

    fun calibrateSittingPosture() {
        val elapsedRealtimeMs = SystemClock.elapsedRealtime()
        val result = analysisCoordinator.captureCalibration(
            nowEpochMs = System.currentTimeMillis(),
        )
        analysisState.value = result.snapshot.toUiState(
            updatedAtElapsedMs = elapsedRealtimeMs,
            calibrationMessage = result.message,
        )
        refreshModelingState(
            bleState = controller.uiState.value,
            analysis = analysisState.value,
        )
    }

    fun setUpperBackSensor(
        sensorId: Int?,
    ) {
        val elapsedRealtimeMs = SystemClock.elapsedRealtime()
        val snapshot = analysisCoordinator.setUpperBackSensor(
            sensorId = sensorId,
            nowEpochMs = System.currentTimeMillis(),
        )
        analysisState.value = snapshot.toUiState(
            updatedAtElapsedMs = elapsedRealtimeMs,
            calibrationMessage = snapshot.calibrationMessage,
        )
        refreshModelingState(
            bleState = controller.uiState.value,
            analysis = analysisState.value,
        )
    }

    fun setLowerBackSensor(
        sensorId: Int?,
    ) {
        val elapsedRealtimeMs = SystemClock.elapsedRealtime()
        val snapshot = analysisCoordinator.setLowerBackSensor(
            sensorId = sensorId,
            nowEpochMs = System.currentTimeMillis(),
        )
        analysisState.value = snapshot.toUiState(
            updatedAtElapsedMs = elapsedRealtimeMs,
            calibrationMessage = snapshot.calibrationMessage,
        )
        refreshModelingState(
            bleState = controller.uiState.value,
            analysis = analysisState.value,
        )
    }

    fun startRecording() {
        if (recordingState.value.isRecording) {
            return
        }

        viewModelScope.launch {
            try {
                val session = sessionRecorder.startSession(controller.uiState.value.deviceName)
                recordingState.value = RecordingState(
                    isRecording = true,
                    sessionName = session.fileName,
                    sessionPath = session.absolutePath,
                    startedAtElapsedMs = SystemClock.elapsedRealtime(),
                    recordedSampleCount = 0,
                    recordedStatusCount = 0,
                    errorMessage = null,
                )
                lastRecordedSampleReceivedAtElapsedMs = null
                lastRecordedStatusReceivedAtElapsedMs = null
                recordCurrentSnapshots(controller.uiState.value)
                refreshSavedSessions()
            } catch (error: Exception) {
                recordingState.update {
                    it.copy(errorMessage = error.message ?: "Failed to start recording.")
                }
            }
        }
    }

    fun stopRecording() {
        if (!recordingState.value.isRecording) {
            return
        }

        viewModelScope.launch {
            try {
                sessionRecorder.stopSession(reason = "user_stopped")
                recordingState.update {
                    it.copy(
                        isRecording = false,
                        startedAtElapsedMs = null,
                        errorMessage = null,
                    )
                }
                lastRecordedSampleReceivedAtElapsedMs = null
                lastRecordedStatusReceivedAtElapsedMs = null
                refreshSavedSessions()
            } catch (error: Exception) {
                recordingState.update {
                    it.copy(errorMessage = error.message ?: "Failed to stop recording.")
                }
            }
        }
    }

    override fun onCleared() {
        if (recordingState.value.isRecording) {
            runBlocking {
                sessionRecorder.stopSession(reason = "viewmodel_cleared")
            }
        }
        resetAnalysis()
        controller.stop()
        super.onCleared()
    }

    companion object {
        fun factory(applicationContext: Context): ViewModelProvider.Factory =
            viewModelFactory {
                initializer {
                    AggregatorViewModel(applicationContext)
                }
            }
    }

    private suspend fun recordIfNeeded(bleState: AggregatorUiState) {
        if (!recordingState.value.isRecording) {
            return
        }

        val sampleTimestamp = bleState.lastSampleReceivedAtElapsedMs
        if (bleState.latestSample != null &&
            sampleTimestamp != null &&
            sampleTimestamp != lastRecordedSampleReceivedAtElapsedMs
        ) {
            val wroteSample = sessionRecorder.appendSample(
                recordedAtEpochMs = System.currentTimeMillis(),
                receivedAtElapsedMs = sampleTimestamp,
                sample = bleState.latestSample,
            )
            if (wroteSample) {
                lastRecordedSampleReceivedAtElapsedMs = sampleTimestamp
                recordingState.update {
                    it.copy(
                        recordedSampleCount = it.recordedSampleCount + 1,
                        errorMessage = null,
                    )
                }
            }
        }

        val statusTimestamp = bleState.lastStatusReceivedAtElapsedMs
        if (bleState.networkStatus != null &&
            statusTimestamp != null &&
            statusTimestamp != lastRecordedStatusReceivedAtElapsedMs
        ) {
            val wroteStatus = sessionRecorder.appendNetworkStatus(
                recordedAtEpochMs = System.currentTimeMillis(),
                receivedAtElapsedMs = statusTimestamp,
                status = bleState.networkStatus,
            )
            if (wroteStatus) {
                lastRecordedStatusReceivedAtElapsedMs = statusTimestamp
                recordingState.update {
                    it.copy(
                        recordedStatusCount = it.recordedStatusCount + 1,
                        errorMessage = null,
                    )
                }
            }
        }
    }

    private suspend fun recordCurrentSnapshots(bleState: AggregatorUiState) {
        recordIfNeeded(bleState)
    }

    private fun analyzeIfNeeded(bleState: AggregatorUiState) {
        val statusTimestamp = bleState.lastStatusReceivedAtElapsedMs
        if (bleState.networkStatus != null &&
            statusTimestamp != null &&
            statusTimestamp != lastAnalyzedStatusReceivedAtElapsedMs
        ) {
            val snapshot = analysisCoordinator.onNetworkStatus(
                receivedAtEpochMs = System.currentTimeMillis(),
                status = bleState.networkStatus,
            )
            lastAnalyzedStatusReceivedAtElapsedMs = statusTimestamp
            analysisState.value = snapshot.toUiState(
                updatedAtElapsedMs = statusTimestamp,
                calibrationMessage = analysisState.value.calibrationMessage,
            )
        }

        val sampleTimestamp = bleState.lastSampleReceivedAtElapsedMs
        if (bleState.latestSample != null &&
            sampleTimestamp != null &&
            sampleTimestamp != lastAnalyzedSampleReceivedAtElapsedMs
        ) {
            val snapshot = analysisCoordinator.onSample(
                receivedAtEpochMs = System.currentTimeMillis(),
                sample = bleState.latestSample,
            )
            lastAnalyzedSampleReceivedAtElapsedMs = sampleTimestamp
            analysisState.value = snapshot.toUiState(
                updatedAtElapsedMs = sampleTimestamp,
                calibrationMessage = analysisState.value.calibrationMessage,
            )
        }
    }

    private suspend fun refreshSavedSessions() {
        recordingState.update {
            it.copy(savedSessions = sessionRecorder.listSessions())
        }
    }

    private fun resetAnalysis() {
        analysisCoordinator.clear()
        analysisState.value = AnalysisUiState()
        latestSamplesBySensor.clear()
        modelingState.value = ModelingUiState()
        lastAnalyzedSampleReceivedAtElapsedMs = null
        lastAnalyzedStatusReceivedAtElapsedMs = null
    }

    private fun updateModelingState(bleState: AggregatorUiState) {
        val sample = bleState.latestSample
        val sampleTimestamp = bleState.lastSampleReceivedAtElapsedMs
        if (sample != null && sampleTimestamp != null) {
            latestSamplesBySensor[sample.sensorId] = TimedSensorSample(
                sample = sample,
                receivedAtElapsedMs = sampleTimestamp,
            )
        }
        refreshModelingState(
            bleState = bleState,
            analysis = analysisState.value,
        )
    }

    private fun refreshModelingState(
        bleState: AggregatorUiState,
        analysis: AnalysisUiState,
    ) {
        val liveSensorIds = bleState.networkStatus?.sensors
            ?.map { it.sensorId }
            ?.toSet()
            .orEmpty()
        val placementBySensorId = analysis.sensorAssignments.associate { assignment ->
            assignment.sensorId to assignment.placement
        }
        val sensorIds = (liveSensorIds + analysis.availableSensorIds + latestSamplesBySensor.keys)
            .toSet()
            .sorted()

        val nodes = sensorIds.map { sensorId ->
            val timedSample = latestSamplesBySensor[sensorId]
            val pose = timedSample?.let { NodePoseMath.fromSample(it.sample) }
            ModeledNodeUiState(
                sensorId = sensorId,
                placement = placementBySensorId[sensorId],
                isLiveInNetworkStatus = sensorId in liveSensorIds,
                seq = timedSample?.sample?.seq,
                pitchDeg = pose?.pitchDeg,
                rollDeg = pose?.rollDeg,
                gravityVector = pose?.gravityVector?.let { gravity ->
                    ModeledGravityVector(
                        x = gravity.x,
                        y = gravity.y,
                        z = gravity.z,
                    )
                },
                lastSampleReceivedAtElapsedMs = timedSample?.receivedAtElapsedMs,
            )
        }

        val statusMessage = when {
            nodes.isEmpty() -> "Waiting for network status and live IMU samples before the modeling scene can render."
            nodes.any { it.gravityVector != null } -> "Orientation-only node scene from the latest per-sensor IMU samples. Positions are schematic until a full spatial model exists."
            else -> "Sensors are known, but the app is still waiting for per-sensor IMU samples to estimate orientation."
        }

        modelingState.value = ModelingUiState(
            nodes = nodes,
            lastUpdatedAtElapsedMs = bleState.lastSampleReceivedAtElapsedMs ?: bleState.lastStatusReceivedAtElapsedMs,
            statusMessage = statusMessage,
        )
    }
}

private data class RecordingState(
    val isRecording: Boolean = false,
    val sessionName: String? = null,
    val sessionPath: String? = null,
    val startedAtElapsedMs: Long? = null,
    val recordedSampleCount: Int = 0,
    val recordedStatusCount: Int = 0,
    val savedSessions: List<SavedSessionSummary> = emptyList(),
    val errorMessage: String? = null,
)

private data class TimedSensorSample(
    val sample: com.example.seniordesignmobileapp.model.ImuSample,
    val receivedAtElapsedMs: Long,
)

private fun com.example.seniordesignmobileapp.domain.PostureAnalysisSnapshot.toUiState(
    updatedAtElapsedMs: Long,
    calibrationMessage: String?,
): AnalysisUiState =
    AnalysisUiState(
        config = config,
        sensorAssignments = sensorAssignments,
        availableSensorIds = availableSensorIds,
        manualUpperBackSensorId = manualUpperBackSensorId,
        manualLowerBackSensorId = manualLowerBackSensorId,
        expectedSensors = expectedSensors,
        expectedSensorsInferred = expectedSensorsInferred,
        sittingCalibration = sittingCalibration,
        windowSummary = windowSummary,
        latestResult = latestResult,
        lastUpdatedAtElapsedMs = updatedAtElapsedMs,
        calibrationMessage = calibrationMessage ?: this.calibrationMessage,
        statusMessage = statusMessage,
    )

File: mobile/ios/FH07Mobile/ViewModel/AggregatorViewModel.swift
import Foundation
import Combine

@MainActor
final class AggregatorViewModel: ObservableObject {
    @Published private(set) var uiState = AggregatorUiState()

    private let controller = BleAggregatorController()
    private let sessionRecorder = SessionRecorder()
    private let analysisCoordinator = PostureAnalysisCoordinator()

    private var recordingState = RecordingStateHolder()
    private var analysisState = AnalysisUiState()
    private var modelingState = ModelingUiState()

    private var latestSamplesBySensor: [Int: TimedSensorSample] = [:]
    private var latestSampleOrder: [Int] = []
    private var lastRecordedSampleReceivedAtElapsedMs: Int64? = nil
    private var lastRecordedStatusReceivedAtElapsedMs: Int64? = nil
    private var lastAnalyzedSampleReceivedAtElapsedMs: Int64? = nil
    private var lastAnalyzedStatusReceivedAtElapsedMs: Int64? = nil

    private var cancellables: Set<AnyCancellable> = []

    init() {
        controller.$uiState
            .receive(on: DispatchQueue.main)
            .sink { [weak self] bleState in
                guard let self = self else { return }
                Task { await self.handleBleStateUpdate(bleState) }
            }
            .store(in: &cancellables)

        Task { await refreshSavedSessions() }
        republishUiState(controller.uiState)
    }

    func onBlePermissionsChanged(granted: Bool) {
        if granted {
            controller.start()
        } else {
            resetAnalysis()
            controller.stop()
        }
    }

    func reconnect() {
        resetAnalysis()
        controller.restart()
    }

    func calibrateSittingPosture() {
        let elapsedRealtimeMsValue = elapsedRealtimeMs()
        let result = analysisCoordinator.captureCalibration(
            nowEpochMs: Int64(Date().timeIntervalSince1970 * 1000)
        )
        analysisState = toUiState(
            snapshot: result.snapshot,
            updatedAtElapsedMs: elapsedRealtimeMsValue,
            calibrationMessage: result.message
        )
        refreshModelingState(bleState: controller.uiState, analysis: analysisState)
        republishUiState(controller.uiState)
    }

    func setUpperBackSensor(_ sensorId: Int?) {
        let elapsedRealtimeMsValue = elapsedRealtimeMs()
        let snapshot = analysisCoordinator.setUpperBackSensor(
            sensorId: sensorId,
            nowEpochMs: Int64(Date().timeIntervalSince1970 * 1000)
        )
        analysisState = toUiState(
            snapshot: snapshot,
            updatedAtElapsedMs: elapsedRealtimeMsValue,
            calibrationMessage: snapshot.calibrationMessage
        )
        refreshModelingState(bleState: controller.uiState, analysis: analysisState)
        republishUiState(controller.uiState)
    }

    func setLowerBackSensor(_ sensorId: Int?) {
        let elapsedRealtimeMsValue = elapsedRealtimeMs()
        let snapshot = analysisCoordinator.setLowerBackSensor(
            sensorId: sensorId,
            nowEpochMs: Int64(Date().timeIntervalSince1970 * 1000)
        )
        analysisState = toUiState(
            snapshot: snapshot,
            updatedAtElapsedMs: elapsedRealtimeMsValue,
            calibrationMessage: snapshot.calibrationMessage
        )
        refreshModelingState(bleState: controller.uiState, analysis: analysisState)
        republishUiState(controller.uiState)
    }

    func startRecording() {
        if recordingState.isRecording { return }

        Task {
            do {
                let deviceName = controller.uiState.deviceName
                let session = try await sessionRecorder.startSession(deviceName: deviceName)
                recordingState = RecordingStateHolder(
                    isRecording: true,
                    sessionName: session.fileName,
                    sessionPath: session.absolutePath,
                    startedAtElapsedMs: elapsedRealtimeMs(),
                    recordedSampleCount: 0,
                    recordedStatusCount: 0,
                    savedSessions: recordingState.savedSessions,
                    errorMessage: nil
                )
                lastRecordedSampleReceivedAtElapsedMs = nil
                lastRecordedStatusReceivedAtElapsedMs = nil
                await recordIfNeeded(bleState: controller.uiState)
                await refreshSavedSessions()
                republishUiState(controller.uiState)
            } catch {
                recordingState.errorMessage = error.localizedDescription
                republishUiState(controller.uiState)
            }
        }
    }

    func stopRecording() {
        if !recordingState.isRecording { return }

        Task {
            _ = await sessionRecorder.stopSession(reason: "user_stopped")
            recordingState.isRecording = false
            recordingState.startedAtElapsedMs = nil
            recordingState.errorMessage = nil
            lastRecordedSampleReceivedAtElapsedMs = nil
            lastRecordedStatusReceivedAtElapsedMs = nil
            await refreshSavedSessions()
            republishUiState(controller.uiState)
        }
    }

    deinit {
        // Recording cleanup: the actor is reference-counted; sessions are closed when the recorder deinits.
    }

    // MARK: - State update pipeline

    private func handleBleStateUpdate(_ bleState: AggregatorUiState) async {
        await recordIfNeeded(bleState: bleState)
        analyzeIfNeeded(bleState: bleState)
        updateModelingState(bleState: bleState)
        republishUiState(bleState)
    }

    private func recordIfNeeded(bleState: AggregatorUiState) async {
        guard recordingState.isRecording else { return }

        if let sample = bleState.latestSample,
           let sampleTimestamp = bleState.lastSampleReceivedAtElapsedMs,
           sampleTimestamp != lastRecordedSampleReceivedAtElapsedMs {
            let wrote = await sessionRecorder.appendSample(
                recordedAtEpochMs: Int64(Date().timeIntervalSince1970 * 1000),
                receivedAtElapsedMs: sampleTimestamp,
                sample: sample
            )
            if wrote {
                lastRecordedSampleReceivedAtElapsedMs = sampleTimestamp
                recordingState.recordedSampleCount += 1
                recordingState.errorMessage = nil
            }
        }

        if let status = bleState.networkStatus,
           let statusTimestamp = bleState.lastStatusReceivedAtElapsedMs,
           statusTimestamp != lastRecordedStatusReceivedAtElapsedMs {
            let wrote = await sessionRecorder.appendNetworkStatus(
                recordedAtEpochMs: Int64(Date().timeIntervalSince1970 * 1000),
                receivedAtElapsedMs: statusTimestamp,
                status: status
            )
            if wrote {
                lastRecordedStatusReceivedAtElapsedMs = statusTimestamp
                recordingState.recordedStatusCount += 1
                recordingState.errorMessage = nil
            }
        }
    }

    private func analyzeIfNeeded(bleState: AggregatorUiState) {
        if let status = bleState.networkStatus,
           let statusTimestamp = bleState.lastStatusReceivedAtElapsedMs,
           statusTimestamp != lastAnalyzedStatusReceivedAtElapsedMs {
            let snapshot = analysisCoordinator.onNetworkStatus(
                receivedAtEpochMs: Int64(Date().timeIntervalSince1970 * 1000),
                status: status
            )
            lastAnalyzedStatusReceivedAtElapsedMs = statusTimestamp
            analysisState = toUiState(
                snapshot: snapshot,
                updatedAtElapsedMs: statusTimestamp,
                calibrationMessage: analysisState.calibrationMessage
            )
        }

        if let sample = bleState.latestSample,
           let sampleTimestamp = bleState.lastSampleReceivedAtElapsedMs,
           sampleTimestamp != lastAnalyzedSampleReceivedAtElapsedMs {
            let snapshot = analysisCoordinator.onSample(
                receivedAtEpochMs: Int64(Date().timeIntervalSince1970 * 1000),
                sample: sample
            )
            lastAnalyzedSampleReceivedAtElapsedMs = sampleTimestamp
            analysisState = toUiState(
                snapshot: snapshot,
                updatedAtElapsedMs: sampleTimestamp,
                calibrationMessage: analysisState.calibrationMessage
            )
        }
    }

    private func refreshSavedSessions() async {
        let sessions = await sessionRecorder.listSessions()
        recordingState.savedSessions = sessions
        republishUiState(controller.uiState)
    }

    private func resetAnalysis() {
        analysisCoordinator.clear()
        analysisState = AnalysisUiState()
        latestSamplesBySensor.removeAll()
        latestSampleOrder.removeAll()
        modelingState = ModelingUiState()
        lastAnalyzedSampleReceivedAtElapsedMs = nil
        lastAnalyzedStatusReceivedAtElapsedMs = nil
    }

    private func updateModelingState(bleState: AggregatorUiState) {
        if let sample = bleState.latestSample,
           let sampleTimestamp = bleState.lastSampleReceivedAtElapsedMs {
            if latestSamplesBySensor[sample.sensorId] == nil {
                latestSampleOrder.append(sample.sensorId)
            }
            latestSamplesBySensor[sample.sensorId] = TimedSensorSample(
                sample: sample,
                receivedAtElapsedMs: sampleTimestamp
            )
        }
        refreshModelingState(bleState: bleState, analysis: analysisState)
    }

    private func refreshModelingState(bleState: AggregatorUiState, analysis: AnalysisUiState) {
        let liveSensorIds: Set<Int> = Set(bleState.networkStatus?.sensors.map { $0.sensorId } ?? [])
        var placementBySensorId: [Int: SensorPlacement] = [:]
        for assignment in analysis.sensorAssignments {
            placementBySensorId[assignment.sensorId] = assignment.placement
        }

        var allIds = liveSensorIds
        for id in analysis.availableSensorIds { allIds.insert(id) }
        for id in latestSamplesBySensor.keys { allIds.insert(id) }
        let sensorIds = allIds.sorted()

        let nodes: [ModeledNodeUiState] = sensorIds.map { sensorId in
            let timedSample = latestSamplesBySensor[sensorId]
            let pose: NodePoseEstimate? = timedSample.flatMap { NodePoseMath.fromSample($0.sample) }
            return ModeledNodeUiState(
                sensorId: sensorId,
                placement: placementBySensorId[sensorId],
                isLiveInNetworkStatus: liveSensorIds.contains(sensorId),
                seq: timedSample?.sample.seq,
                pitchDeg: pose?.pitchDeg,
                rollDeg: pose?.rollDeg,
                gravityVector: pose.map { ModeledGravityVector(x: $0.gravityVector.x, y: $0.gravityVector.y, z: $0.gravityVector.z) },
                lastSampleReceivedAtElapsedMs: timedSample?.receivedAtElapsedMs
            )
        }

        let statusMessage: String
        if nodes.isEmpty {
            statusMessage = "Waiting for network status and live IMU samples before the modeling scene can render."
        } else if nodes.contains(where: { $0.gravityVector != nil }) {
            statusMessage = "Orientation-only node scene from the latest per-sensor IMU samples. Positions are schematic until a full spatial model exists."
        } else {
            statusMessage = "Sensors are known, but the app is still waiting for per-sensor IMU samples to estimate orientation."
        }

        modelingState = ModelingUiState(
            nodes: nodes,
            lastUpdatedAtElapsedMs: bleState.lastSampleReceivedAtElapsedMs ?? bleState.lastStatusReceivedAtElapsedMs,
            statusMessage: statusMessage
        )
    }

    private func republishUiState(_ bleState: AggregatorUiState) {
        var next = bleState
        next.isRecording = recordingState.isRecording
        next.recordingSessionName = recordingState.sessionName
        next.recordingSessionPath = recordingState.sessionPath
        next.recordingStartedAtElapsedMs = recordingState.startedAtElapsedMs
        next.recordedSampleCount = recordingState.recordedSampleCount
        next.recordedStatusCount = recordingState.recordedStatusCount
        next.savedSessions = recordingState.savedSessions
        next.recordingErrorMessage = recordingState.errorMessage
        next.analysis = analysisState
        next.modeling = modelingState
        uiState = next
    }

    private func toUiState(
        snapshot: PostureAnalysisSnapshot,
        updatedAtElapsedMs: Int64,
        calibrationMessage: String?
    ) -> AnalysisUiState {
        AnalysisUiState(
            config: snapshot.config,
            sensorAssignments: snapshot.sensorAssignments,
            availableSensorIds: snapshot.availableSensorIds,
            manualUpperBackSensorId: snapshot.manualUpperBackSensorId,
            manualLowerBackSensorId: snapshot.manualLowerBackSensorId,
            expectedSensors: snapshot.expectedSensors,
            expectedSensorsInferred: snapshot.expectedSensorsInferred,
            sittingCalibration: snapshot.sittingCalibration,
            windowSummary: snapshot.windowSummary,
            latestResult: snapshot.latestResult,
            lastUpdatedAtElapsedMs: updatedAtElapsedMs,
            calibrationMessage: calibrationMessage ?? snapshot.calibrationMessage,
            statusMessage: snapshot.statusMessage
        )
    }
}

private struct RecordingStateHolder {
    var isRecording: Bool = false
    var sessionName: String? = nil
    var sessionPath: String? = nil
    var startedAtElapsedMs: Int64? = nil
    var recordedSampleCount: Int = 0
    var recordedStatusCount: Int = 0
    var savedSessions: [SavedSessionSummary] = []
    var errorMessage: String? = nil
}

private struct TimedSensorSample {
    let sample: ImuSample
    let receivedAtElapsedMs: Int64
}
